src/util.rs

```diff
@@ -177,33 +185,42 @@ impl From<io::Error> for BaoValidationError {
 }
 
 /// Validate that the data matches the outboard.
-pub(crate) fn validate_bao(
+pub(crate) fn validate_bao<F: Fn(u64)>(
     hash: Hash,
     data_reader: impl Read + Seek,
     outboard: Bytes,
-    progress: impl Fn(u64),
+    progress: F,
 ) -> result::Result<(), BaoValidationError> {
     let hash = blake3::Hash::from(hash);
-    let outboard_reader = io::Cursor::new(outboard);
-    let progress_reader = ProgressReader::new(data_reader, |p| {
-        if let ProgressReaderUpdate::Progress(x) = p {
-            progress(x)
-        }
-    });
-    let buffered_reader = BufReader::with_capacity(1024 * 1024, progress_reader);
-    let mut decoder = abao::decode::Decoder::new_outboard(buffered_reader, outboard_reader, &hash);
-    // todo: expose chunk group size in abao, so people can allocate good sized buffers
-    let mut buffer = vec![0u8; 1024 * 16 + 4096];
-    loop {
-        match decoder.read(&mut buffer) {
-            Ok(0) => break,
-            Ok(_) => {}
-            Err(err) => {
-                // todo: figure out exactly what went wrong
-                return Err(BaoValidationError::IoError(err));
+    let outboard =
+        bao_tree::outboard::PreOrderMemOutboard::new(hash, IROH_BLOCK_SIZE, outboard.to_vec())
```

Comment:
that `to_vec` call is sad

Reply:
This entire section is sad, but it is going to be changed.

Reply:
Got rid of the to_vec(). But it is really annoying having to write code twice, once for owned and once for borrowed...

Reply:
Cow?

Reply:
it is back